src/main/java/org/wildfly/security/auth/client/ElytronXmlParser.java

```diff
@@ -1007,9 +1028,272 @@ static void parseKeyStoreType(ConfigurationXMLStreamReader reader, final Map<Str
         throw reader.unexpectedDocumentEnd();
     }
 
+    /**
+     * Parse an XML element of type {@code credential-store-reference-type} from an XML reader.
+     *
+     * @param reader the XML stream reader
+     * @param credentialStoresNames set of Credential Store names already defined used for checking of duplicates
+     * @return {@link CredentialStoreReference} corresponding to given XML configuration element
+     * @throws ConfigXMLParseException if the resource failed to be parsed
+     */
+    public static CredentialStoreReference parseCredentialStoreRefType(ConfigurationXMLStreamReader reader, final Map<String, SecurityFactory<CredentialStore>> credentialStoresNames) throws ConfigXMLParseException {
+        final int attributeCount = reader.getAttributeCount();
+        final CredentialStoreReference credentialStoreReference;
+        String store = null;
+        String alias = null;
+        String type = null;
+        String clearText = null;
+        for (int i = 0; i < attributeCount; i ++) {
+            final String attributeNamespace = reader.getAttributeNamespace(i);
+            if (attributeNamespace != null && ! attributeNamespace.isEmpty()) {
+                throw reader.unexpectedAttribute(i);
+            }
+            switch (reader.getAttributeLocalName(i)) {
+                case "store": {
+                    if (store != null) throw reader.unexpectedAttribute(i);
+                    store = reader.getAttributeValue(i);
+                    if (credentialStoresNames != null && !credentialStoresNames.containsKey(store)) {
+                        throw xmlLog.credentialStoreNameNotDefined(reader, store);
+                    }
+                    break;
+                }
+                case "alias": {
+                    if (alias != null) throw reader.unexpectedAttribute(i);
+                    alias = reader.getAttributeValue(i);
+                    break;
+                }
+                case "type": {
+                    if (type != null) throw reader.unexpectedAttribute(i);
+                    type = reader.getAttributeValue(i);
+                    break;
+                }
+                case "clear-text": {
+                    if (clearText != null) throw reader.unexpectedAttribute(i);
+                    clearText = reader.getAttributeValue(i);
+                    break;
+                }
+                default:
+                    throw reader.unexpectedAttribute(i);
+            }
+        }
+        if (store == null) {
+            if (clearText == null) {
+                throw missingAttribute(reader, "clear-text");
+            }
+            credentialStoreReference = new CredentialStoreReference(clearText.toCharArray());
+        } else {
+            if (alias == null) {
+                throw missingAttribute(reader, "alias");
+            }
+            credentialStoreReference = new CredentialStoreReference(store, alias, type);
```

Comment:
In places where we're loading types by name, we must always have the option to specify the module name as well so that in a modular environment, the credential type is loaded from that module.  If no module is given, the current class loader should be used.  There should be at least one existing example of this around here someone.